Add unit test for Run 3 Scouting formats

wddgit opened the pull request:
#### PR description:

Add a new unit test for the Run 3 Scouting data formats. This generates a data file containing Run 3 Scouting objects with known content. Then it reads it. It verifies that when we read it we obtain values that match the known written values for all the data fields in the objects. In particular all containers have content so all contained types are also read.

It also reads the old files in the DataFormats-Scouting data repository which are listed in the shell script. The plan is that each time a Run 3 Scouting data format is modified an additional file will added. The data files required for this PR were generated with releases 12_0_4 (2022 data) and 13_0_3 (2023 data).

There are 10 events with content and vector sizes varying from event to event.

#### PR validation:

This only adds a unit test in DataFormats/Scouting which passes. It shouldn't affect anything else.

Note that the test will only pass if the cms-data/DataFormats-Scouting#1 pull request is merged and that external update is used. Waiting to merge this PR until a couple days after the data PR is merged might avoid some syncing difficulties in IBs and PR tests.

wddgit committed:
Implement Run3 Scouting test, CMSSW_12_4_0 version.

wddgit committed:
Run 3 Scouting Test CMSSW_13_0_3 version

wddgit committed:
Run 3 Scouting format tests, read old versions in unit test

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41834/35730

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @wddgit (W. David Dagenhart) for master.

It involves the following packages:

- DataFormats/Scouting (**core**)


@cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please review it and eventually sign? Thanks.
@missirol, @rovere this is something you requested to watch as well.
@perrotta, @dpiparo, @rappoccio you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

wddgit commented:
please test with cms-data/DataFormats-Scouting#1

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32912/summary.html
**COMMIT**: 084a0b81e8b1d3baad5a6c5d2d28feff78bf7d55
**CMSSW**: CMSSW_13_2_X_2023-05-31-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41834/32912/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @smuzaffar cms-sw/cmssw#41771
 - @smuzaffar cms-sw/cmssw#41764
 - @bsunanda cms-sw/cmssw#41788
 - @aloeliger cms-sw/cmssw#41780
 - @bsunanda cms-sw/cmssw#41791
 - @smorovic cms-sw/cmssw#41814
 - @mmusich cms-sw/cmssw#41816
 - @VinInn cms-sw/cmssw#41820
 - @civanch cms-sw/cmssw#41826
 - @smuzaffar cms-sw/cmssw#41762

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32912/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32912/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 11 lines to the logs
 - Reco comparison results: 8 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3221457
 - DQMHistoTests: Total failures: 10
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3221425
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-05-31-1100+a96aee/57296/triggerResults): no differences found

makortel commented on DataFormats/Scouting/README.md:
```diff
@@ -0,0 +1,10 @@
+* a file written by the same release
+* files written by (some) earlier releases
+
+If the persistent format of any Scouting data format gets changed in the future, please adjust the `TestReadRun3Scouting` and `TestWriteRun3Scouting` modules accordingly. It is important that every member container has some content in this test. Please also add a new file to the [https://github.com/cms-data/DataFormats-Scouting/](https://github.com/cms-data/DataFormats-Scouting/) repository, and  ...
```
I'd update the text with the outcome of the discussion in https://github.com/cms-data/DataFormats-Scouting/pull/1#discussion_r1213456131 . I'd also add here a comment along "if the latest file of Run 3 scouting before the update has not been used in data taking, the file can be deleted".

wddgit replied:
Done.

makortel commented on DataFormats/Scouting/test/TestReadRun3Scouting.cc:
```diff
@@ -0,0 +1,1091 @@
+    // to check that values read from persistent storage match the values
+    // we know were written.
+
+    std::vector<double> expectedCaloJetsValues_;
```
Seems like all of these could be `const`

wddgit replied:
I made all the data members const in both modules.

wddgit commented on DataFormats/Scouting/test/TestRun3ScoutingFormats.sh:
```diff
@@ -0,0 +1,43 @@
+# rename the file it creates.
+#
+# 12_4_0 was chosen because that release was used for 2022 data.
+# 13_0_3 is being used for 2023 data.
```
Done.

wddgit commented on DataFormats/Scouting/test/TestReadRun3Scouting.cc:
```diff
@@ -0,0 +1,1091 @@
+  void TestReadRun3Scouting::analyzeCaloJets(edm::Event const& iEvent) const {
+    if (expectedCaloJetsValues_.size() != 16) {
+      throwWithMessage("analyzeCaloJets, test configuration error, expectedCaloJetsValues must have size 16");
+    }
```
I moved them to the constructor. Thanks.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41834/35750

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - DataFormats/Scouting/test/create_Run3Scouting_test_file.py:
    - Added: 8503fc93fa7b31d169836c95fe276056a6207ec5
    - Modified: 21991a15d200d9277e28c6b0f17f60f59fefe044
    - Deleted: e5ba4b384585e849837b34f6ea011bff8e9d6067

cmsbuild commented:
Pull request #41834 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented:
please test with cms-data/DataFormats-Scouting#1

I think I took care of all the comments. I just squashed them into the last commit. Hopefully all is good now. Thanks for the review. In addition, I renamed create_Run3Scouting_test_file.py to create_Run3Scouting_test_file_cfg.py

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32943/summary.html
**COMMIT**: e5ba4b384585e849837b34f6ea011bff8e9d6067
**CMSSW**: CMSSW_13_2_X_2023-06-01-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41834/32943/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @smorovic cms-sw/cmssw#41837
 - @gartung cms-sw/cmssw#41833
 - @OlivieFranklova cms-sw/cmssw#41589

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32943/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32943/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially removed 2 lines from the logs
 - ROOTFileChecks: Some differences in event products or their sizes found
 - Reco comparison results: 175 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3186957
 - DQMHistoTests: Total failures: 3
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3186932
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 168.456 KiB( 47 files compared)
 - DQMHistoSizes: changed ( 23234.0,... ): 28.076 KiB       HGCAL/HGCalValidator
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-01-1100+a96aee/57317/triggerResults): found differences in 1 / 46 workflows

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-41834/35758

- This PR adds an extra 28KB to repository

- Found files with invalid states:
  - DataFormats/Scouting/test/create_Run3Scouting_test_file.py:
    - Added: 8503fc93fa7b31d169836c95fe276056a6207ec5
    - Modified: 21991a15d200d9277e28c6b0f17f60f59fefe044
    - Deleted: 2844104e1cd225b458b546c2844230377b5a6d09

cmsbuild commented:
Pull request #41834 was updated. @cmsbuild, @smuzaffar, @Dr15Jones, @makortel can you please check and sign again.

wddgit commented:
please test with cms-data/DataFormats-Scouting#1

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32953/summary.html
**COMMIT**: 2844104e1cd225b458b546c2844230377b5a6d09
**CMSSW**: CMSSW_13_2_X_2023-06-02-1100/el8_amd64_gcc11
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/41834/32953/install.sh` to create a dev area with all the needed externals and cmssw changes.



The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - @eyigitba cms-sw/cmssw#41835

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32953/git-recent-commits.json
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-a96aee/32953/git-merge-result


## Comparison Summary



**Summary**:
 - You potentially added 16 lines to the logs
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 48
 - DQMHistoTests: Total histograms compared: 3221457
 - DQMHistoTests: Total failures: 6
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3221429
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 47 files compared)
 - Checked 207 log files, 159 edm output root files, 48 DQM output files
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_2_X_2023-06-02-1100+a96aee/57323/triggerResults): no differences found

makortel commented:
+core

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @perrotta, @dpiparo, @rappoccio (and backports should be raised in the release meeting by the corresponding L2)

perrotta commented on DataFormats/Scouting/README.md:
```diff
@@ -0,0 +1,12 @@
+* a file written by the same release
+* files written by (some) earlier releases
+
+If the persistent format of any Scouting data format gets changed in the future, please adjust the `TestReadRun3Scouting` and `TestWriteRun3Scouting` modules accordingly. It is important that every member container has some content in this test. Please also add a new file to the [https://github.com/cms-data/DataFormats-Scouting/](https://github.com/cms-data/DataFormats-Scouting/) repository, and  ...
```
It wouldn't be bad to also rearrange classes_def.xml in order to have the classes listed there in alphabetical order, in order to avoid some head hache to who's trying to associate the versions in the name of the data file to the classes.

makortel replied:
Thanks @perrotta for the suggestion, I agree a matching order in the `classes_def.xml` would help. @wddgit Could you take care of it (either as part of the Run2 scouting PR, or in a separate PR)?

wddgit replied:
I'll include it in the Run 2 Scouting PR which I am currently implementing. A good idea.

perrotta commented:
+1
